tasks/mesos.yml

```diff
@@ -19,62 +19,93 @@
-    - ansible_service_mgr != 'systemd'
-  notify:
-    - Restart mesos-slave
-
-- name: Check if systemd exists
-  stat: path=/lib/systemd/system/
-  register: systemd_check
+  when: '{{ "slave" in mesos_install_mode and upstart_check.stat.exists }}'
+  register: upstartdropinslave
 
 - name: Mesos master systemd script
-  template: src=mesos-master.service.j2 dest=/lib/systemd/system/mesos-master.service
-  when: (mesos_install_mode == "master" or mesos_install_mode == "master-slave") and systemd_check.stat.exists == true
-  notify:
-    - Reload daemon
-    - Restart mesos-master
+  template: src=mesos-master.service.j2 dest=/etc/systemd/system/mesos-master.service
+  when: '{{ "master" in mesos_install_mode and systemd_check.stat.exists }}'
+  register: systemdunitmaster
 
 - name: Mesos slave systemd script
-  template: src=mesos-slave.service.j2 dest=/lib/systemd/system/mesos-slave.service
-  when: (mesos_install_mode == "slave" or mesos_install_mode == "master-slave") and systemd_check.stat.exists == true
-  notify:
-    - Reload daemon
-    - Restart mesos-slave
+  template: src=mesos-slave.service.j2 dest=/etc/systemd/system/mesos-slave.service
+  when: '{{ "slave" in mesos_install_mode and systemd_check.stat.exists }}'
+  register: systemdunitslave
+
+# While waiting for https://github.com/ansible/ansible-modules-core/issues/191
+- name: Reload systemd
+  command: systemctl --system daemon-reload
+  when: '{{ ansible_service_mgr == "systemd" and (systemdunitmaster.changed|default(False) or systemdunitslave.changed|default(False)) }}'
 
 - name: Enable/Disable mesos-master service
-  service: name=mesos-master enabled="{{ (mesos_install_mode == 'slave') | ternary('no', 'yes') }}"
-  when: systemd_check.stat.exists == true
+  service:
+    name: mesos-master
+    enabled: yes
+  when: '{{ "master" in mesos_install_mode }}'
 
 - name: Enable/Disable mesos-slave service
-  service: name=mesos-slave enabled="{{ (mesos_install_mode == 'master') | ternary('no', 'yes') }}"
-  when: systemd_check.stat.exists == true
+  service:
+    name: mesos-slave
+    enabled: yes
+  when: '{{ "slave" in mesos_install_mode }}'
+
+- set_fact:
+    mesos_master_config_updated: '{{    default_config.changed|default(False)
+                                     or master_config.changed|default(False)
+                                     or (upstartdropinmaster.changed|default(False) and ansible_service_mgr == "upstart")
+                                     or (systemdunitmaster.changed|default(False)   and ansible_service_mgr == "systemd") }}'
+    mesos_slave_config_updated:  '{{    default_config.changed|default(False)
+                                     or slave_config.changed|default(False)
+                                     or (upstartdropinslave.changed|default(False)  and ansible_service_mgr == "upstart")
+                                     or (systemdunitslave.changed|default(False)    and ansible_service_mgr == "systemd") }}'
```

Comment:
won't accessing '.changed' fail on `systemdunitslave` if that var does not exist?  (as the var won't be register'ed for all cases .. also think about allowing the role to run with tags).
idem for the other vars here